content/posts/numpy/numpy-rng/index.md

```diff
@@ -0,0 +1,136 @@
+
+## Random number generation with NumPy
+When you import `numpy` in your python script a RNG is created behind the scenes. This RNG is the one used when you generate a new random value using a function such as `np.random.random`. I will here refer to this RNG as the global NumPy RNG.
+
+Although not recommended, it is a common practice to reset the seed of this global RNG at the beginning of a script using the `np.random.seed` function. Fixing the seed at the beginning ensures that the script is reproducible: the same values and results will be produced each time you run it. However, although sometimes convenient, using the global NumPy RNG is considered a bad practice. A simple reason is that using global variables can lead to undesired side effects. For instance one might use `np.random.random` without knowing that the seed of the global RNG was set somewhere else in the codebase. Quoting the [Numpy Enhancement Proposal (NEP) 19](https://numpy.org/neps/nep-0019-rng-policy.html) by Robert Kern about the NumPy RNG policy:
+
+> The implicit global RandomState behind the `np.random.*` convenience functions can cause problems, especially when threads or other forms of concurrency are involved. Global state is always problematic. We categorically recommend avoiding using the convenience functions when reproducibility is involved. [...] The preferred best practice for getting reproducible pseudorandom numbers is to instantiate a generator object with a seed and pass it around.
+
+In short:
+* Instead of using `np.random.seed`, which reseeds the already created global NumPy RNG and then using `np.random.*` functions you should create a new RNG.
+* You should create one RNG at the beginning of your script (with a seed if you want reproducibility) and use this RNG in the rest of your script.
+
+To create a new RNG you can use the `default_rng` function as illustrated in the [introduction of the random module documentation](https://numpy.org/doc/stable/reference/random/index.html):
+
+```python
+import numpy as np
+
+rng = np.random.default_rng()
+rng.random()
+```
+
+If you want to use a seed for reproducibility, [Numpy documentation](https://numpy.org/doc/stable/reference/random/index.html#quick-start) recommends using a very large and unique number to have a different seed than anyone else. Another reason is that using a small number for the seed could lead to a bad initialization of the RNG, although `default_rng` now takes care of creating a good initialization from a small number. To obtain a good unique seed you can rely on the [secrets module](https://docs.python.org/3/library/secrets.html). The following code will return a 128 bit number that you can then use as a seed:
```

Comment:
Maybe worth explaining what a seed is somewhere higher up?

Reply:
not sure I want to enter into this, I suspect the reader to be familiar with seeding and this is a guide on how to do this properly. Or I would need to write a small paragraph in the intro explaining a bit how an RNG works.

Reply:
I think I will add a small paragraph in the intro to clarify all this.